### PR TITLE
Convert Vector to LocalVector in animation system

### PR DESCRIPTION
**Pull request by YYF233333**
Implement [godotengine/godot-proposals#10594](https://github.com/godotengine/godot-proposals/issues/10594)

convert all `Vector` I found in `animation_tree.h` and `animation_blend_tree.h` except:
- `connections` : they involve a copy which may benifit from COW
- `get_editable_animation_list` : seems like an interface

Tested with godot-benchmark, no noticeable performance change. Is this change acceptable as a codestyle improvement?

Build with `scons target=template_release production=yes`, run with `.\godot.windows.template_release.x86_64.console.exe .\godot-benchmarks\project.godot -- --run-benchmarks --include-benchmarks="animation/animated_models/*"`

Test Result:
```
Master Branch:
{"engine":{"version":"v4.4.dev.custom_build","version_hash":"e3213aaef5e0e72b8272e65d989d3d8222be17ca"},"system":{"cpu_architecture":"x86_64","cpu_name":"12th Gen Intel(R) Core(TM) i5-12400","os":"Windows","cpu_count":12}}
{"benchmarks":[{"name":"Animation Blend Tree 100","category":"Animation > Animated Models","results":{"render_cpu":6.455,"render_gpu":64.95,"time":0.036}},{"name":"Animation State Machine 1000","category":"Animation > Animated Models","results":{"render_cpu":6.371,"render_gpu":64.87,"time":0.017}}]}
{"benchmarks":[{"name":"Animation Blend Tree 100","category":"Animation > Animated Models","results":{"render_cpu":6.462,"render_gpu":64.97,"time":0.023}},{"name":"Animation State Machine 1000","category":"Animation > Animated Models","results":{"render_cpu":6.378,"render_gpu":64.82,"time":0.014}}]}
{"benchmarks":[{"name":"Animation Blend Tree 100","category":"Animation > Animated Models","results":{"render_cpu":6.457,"render_gpu":65.08,"time":0.023}},{"name":"Animation State Machine 1000","category":"Animation > Animated Models","results":{"render_cpu":6.313,"render_gpu":64.85,"time":0.017}}]}
{"benchmarks":[{"name":"Animation Blend Tree 100","category":"Animation > Animated Models","results":{"render_cpu":6.451,"render_gpu":65.03,"time":0.032}},{"name":"Animation State Machine 1000","category":"Animation > Animated Models","results":{"render_cpu":6.358,"render_gpu":64.81,"time":0.015}}]}
{"benchmarks":[{"name":"Animation Blend Tree 100","category":"Animation > Animated Models","results":{"render_cpu":6.482,"render_gpu":64.93,"time":0.026}},{"name":"Animation State Machine 1000","category":"Animation > Animated Models","results":{"render_cpu":6.369,"render_gpu":64.82,"time":0.016}}]}

This PR:
{"engine":{"version_hash":"5d3d68dd405a13a4de63b8c39c000fc67bb5dea8","version":"v4.4.dev.custom_build"},"system":{"os":"Windows","cpu_count":12,"cpu_name":"12th Gen Intel(R) Core(TM) i5-12400","cpu_architecture":"x86_64"}}
{"benchmarks":[{"name":"Animation Blend Tree 100","category":"Animation > Animated Models","results":{"render_cpu":6.599,"render_gpu":65.46,"time":0.028}},{"name":"Animation State Machine 1000","category":"Animation > Animated Models","results":{"render_cpu":6.37,"render_gpu":65.55,"time":0.016}}]}
{"benchmarks":[{"name":"Animation Blend Tree 100","category":"Animation > Animated Models","results":{"render_cpu":6.562,"render_gpu":65.04,"time":0.022}},{"name":"Animation State Machine 1000","category":"Animation > Animated Models","results":{"render_cpu":6.434,"render_gpu":64.74,"time":0.026}}]}
{"benchmarks":[{"name":"Animation Blend Tree 100","category":"Animation > Animated Models","results":{"render_cpu":6.562,"render_gpu":66.36,"time":0.024}},{"name":"Animation State Machine 1000","category":"Animation > Animated Models","results":{"render_cpu":6.427,"render_gpu":64.78,"time":0.029}}]}
{"benchmarks":[{"name":"Animation Blend Tree 100","category":"Animation > Animated Models","results":{"render_cpu":6.236,"render_gpu":63.03,"time":0.023}},{"name":"Animation State Machine 1000","category":"Animation > Animated Models","results":{"render_cpu":6.425,"render_gpu":62.2,"time":0.017}}]}
{"benchmarks":[{"name":"Animation Blend Tree 100","category":"Animation > Animated Models","results":{"render_cpu":6.927,"render_gpu":65.25,"time":0.025}},{"name":"Animation State Machine 1000","category":"Animation > Animated Models","results":{"render_cpu":6.874,"render_gpu":65.07,"time":0.017}}]}

# Animation Blend Tree 100 is changed to Animation Blend Tree 1000 because it takes too little time
```

### CODE REVIEW

**Comment by YYF233333** [comment]
oops, I forget to check dev_mode, now tested locally.

**Comment by fire** [comment]
I didn't see any major problems, but I haven't tested.

File: scene/animation/animation_tree.cpp [hunk 1]
```diff
@@ -147,7 +147,7 @@ AnimationTree *AnimationNode::get_animation_tree() const {
 }
 
 AnimationNode::NodeTimeInfo AnimationNode::blend_input(int p_input, AnimationMixer::PlaybackInfo p_playback_info, FilterAction p_filter, bool p_sync, bool p_test_only) {
-	ERR_FAIL_INDEX_V(p_input, inputs.size(), NodeTimeInfo());
+	ERR_FAIL_INDEX_V(p_input, (int64_t)inputs.size(), NodeTimeInfo());
```
Comment (fire) [inline scene/animation/animation_tree.cpp:150]: This is odd isn't there a 64 bit test?

File: scene/animation/animation_tree.cpp [hunk 2]
```diff
@@ -167,12 +167,12 @@ AnimationNode::NodeTimeInfo AnimationNode::blend_input(int p_input, AnimationMix
...
 
 	real_t activity = 0.0;
-	Vector<AnimationTree::Activity> *activity_ptr = process_state->tree->input_activity_map.getptr(node_state.base_path);
+	LocalVector<AnimationTree::Activity> *activity_ptr = process_state->tree->input_activity_map.getptr(node_state.base_path);
 	NodeTimeInfo nti = _blend_node(node, node_name, nullptr, p_playback_info, p_filter, p_sync, p_test_only, &activity);
 
-	if (activity_ptr && p_input < activity_ptr->size()) {
-		activity_ptr->write[p_input].last_pass = process_state->last_pass;
-		activity_ptr->write[p_input].activity = activity;
+	if (activity_ptr && p_input < (int64_t)activity_ptr->size()) {
```
Comment (fire) [inline scene/animation/animation_tree.cpp:173]: Same here.

File: scene/animation/animation_tree.cpp (hunk 1 above)
Comment (YYF233333) [reply scene/animation/animation_tree.cpp:150]: I can't find any 64 bit variation of the macro, not very sure what you mean. The conversion is to supress compiler warning and be aligned with `Vector`'s `size()` (it does this internally and return size in `int64_t`)

**Comment by YYF233333** [comment]
Is this PR still needed? It has been no feedback for several months. I'm going to close this if no body want.

**Review by SaracenOne — APPROVED** [review]
Hello, sorry for the lack of feedback. I have been using this in testing in a custom build for a weeks now and while there could be some nuances of this new approach I may not have fully evaluated the implications of, I haven't noticed any obvious issues with the animation system which seem related to it. I'll approve it for merging, but I'll also ping @TokageItLab in case you have any thoughts on it.

File: scene/animation/animation_tree.cpp [hunk 3]
```diff
@@ -181,12 +181,12 @@ AnimationNode::NodeTimeInfo AnimationNode::blend_input(int p_input, AnimationMix
...
+		activity_ptr->ptr()[p_input].last_pass = process_state->last_pass;
+		activity_ptr->ptr()[p_input].activity = activity;
```
Comment (AThousandShips) [inline scene/animation/animation_tree.cpp:189]:
```suggestion
		(*activity_ptr)[p_input].last_pass = process_state->last_pass;
		(*activity_ptr)[p_input].activity = activity;
```
Cleaner IMO

**Review by Repiteo — APPROVED** [review]
LGTM, just be sure to squash your commits first

**Comment by Repiteo** [comment]
Thanks!

**Pull request merged**